docs(services): update to latest indexer (#2063)
latticexyz/mud: docs/pages/services/indexer.mdx

| Optional                  | RPC_WS_URL      | The URL to access the blockchain using WebSocket                                                         |
| Optional                  | START_BLOCK     | The block to start indexing from. The block in which the `World` contract was deployed is a good choice. | 1                                                         |
| Optional                  | DEBUG=mud:\*    | Turn on debugging                                                                                        |                                                           |
| Optional                  | STORE_ADDRESS   | Only index tables from this `World`                                                                      |
| Optional, only for SQLite | SQLITE_FILENAME | Name of database                                                                                         | `indexer.db`                                              |
| Required for PostgreSQL   | DATABASE_URL    | URL for the database, of the form `postgres://<host>/<database>`                                         |

> ### Schemaful indexing to SQLite via pnpm

Reviewer: ```suggestion
### Schemaful indexing to SQLite via npx
```

To run the indexer directly on your computer using SQLite:

1. Start a `World` to index.
   An easy way to do this is to [use a TypeScript template](/templates/typescript/getting-started) in a separate command line window.

1. Set `RPC_HTTP_URL`.

   ```bash copy
   export RPC_HTTP_URL=http://127.0.0.1:8545
   ```

1. Run the indexer.
   If necessary, install it first.

   ```bash copy
   npx -y -p @latticexyz/store-indexer@next sqlite-indexer
   ```

**Note:** The `indexer.db` is persistent if you stop and restart the indexer.
If that is not the desired behavior (for example, because you restarted the blockchain itself), delete it before starting the indexer.

> ### Schemaless indexing to PostgreSQL via pnpm

Reviewer: ```suggestion
### Schemaless indexing to PostgreSQL via npx
```

This allows it to index the data of all tables of all MUD Worlds on a chain efficiently, but querying for data is limited to filters on `address`, `tableId`, and the record's key.
Since the record's `data` is stored as encoded bytes blob, it's harder to query based on it in SQL.
The main purpose of this variant of the indexer is to efficiently hydrate a MUD client, which decodes the data.

1. Start a `World` to index.
   An easy way to do this is to [use a TypeScript template](/templates/typescript/getting-started) in a separate command line window.

1. Set the environment variables.

   ```bash copy
   export RPC_HTTP_URL=http://127.0.0.1:8545
   export DATABASE_URL=postgres://127.0.0.1/postgres
   ```

1. Run the indexer.
   Install it first if necessary.

   ```bash copy
   npx -y -p @latticexyz/store-indexer@next postgres-indexer
   ```

1. Open a separate command line.
   In it, specify the database and run the indexer frontend, which is responsible for serving the data to the client.

   ```bash copy
   export DATABASE_URL=postgres://127.0.0.1/postgres
   npx -y -p @latticexyz/store-indexer@next postgres-frontend
   ```

> ### Schemaful indexing with PostgreSQL via pnpm

Reviewer: ```suggestion
### Schemaful indexing with PostgreSQL via npx
```

| Needed                    | RPC_HTTP_URL    | The URL to access the blockchain using HTTP                                                              | http://host.docker.internal:8545 (when running in Docker) |
| Optional                  | RPC_WS_URL      | The URL to access the blockchain using WebSocket                                                         |
| Optional                  | START_BLOCK     | The block to start indexing from. The block in which the `World` contract was deployed is a good choice. | 1                                                         |
| Optional                  | DEBUG=mud:\*    | Turn on debugging                                                                                        |                                                           |
| Optional, only for SQLite | SQLITE_FILENAME | Name of database                                                                                         | `indexer.db`                                              |
| Required for PostgreSQL   | DATABASE_URL    | URL for the database, of the form `postgres://<host>/<database>`                                         |

> ### Running directly with SQLite

Reviewer: ```suggestion
### Schemaful indexing to SQLite via pnpm
```

To run the indexer directly on your computer using SQLite:

1. Start a `World` to index.
   An easy way to do this is to [use a TypeScript template](/templates/typescript/getting-started) in a separate command line window.

1. Install the indexer.

   ```bash copy
   pnpm install @latticexyz/store-indexer@next
   ```

1. Set `RPC_HTTP_URL` and run the indexer.

   ```bash copy
   export RPC_HTTP_URL=http://127.0.0.1:8545
   pnpm sqlite-indexer
   ```

**Note:** The `indexer.db` is persistent if you stop and restart the indexer.
If that is not the desired behavior (for example, because you restarted the blockchain itself), delete it before starting the indexer.

> ### Running directly with PostgreSQL storage

Reviewer: ```suggestion
### Schemaless indexing to PostgreSQL via pnpm
```

> This is the method to use PostgreSQL as merely a data storage solution.
> It has better performance, but the format of the data in PostgreSQL is opaque, you have to use our front end to understand the data.

Reviewer: ```suggestion
The schemaless indexer stores MUD table records into a single monolithic table.
This allows it to index the data of all tables of all MUD Worlds on a chain efficiently, but querying for data is limited to filters on `address`, `tableId`, and the record's key.
Since the record's `data` is stored as encoded bytes blob, it's harder to query based on it in SQL.
The main purpose of this variant of the indexer is to efficiently hydrate a MUD client, which decodes the data.
```

1. Start a `World` to index.
   An easy way to do this is to [use a TypeScript template](/templates/typescript/getting-started) in a separate command line window.

1. Install the indexer.

   ```bash copy
   pnpm install @latticexyz/store-indexer@next

>    ```

Reviewer: does it work if we run it via `npx` instead of installing locally?

Author: Yes. Verified and changed.

1. Set the environment variables and start the indexer.

   ```bash copy
   export RPC_HTTP_URL=http://127.0.0.1:8545
   export DATABASE_URL=postgres://127.0.0.1/postgres
   pnpm postgres-indexer
   ```

> 1. Run the front end, which is responsible for serving the data, in a separate command-line window.

Reviewer: ```suggestion
1. Run the indexer frontend, which is responsible for serving the data to the client, in a separate command-line window.
```

   ```bash copy
   export DATABASE_URL=postgres://127.0.0.1/postgres
   pnpm postgres-frontend
   ```

> ### Running directly with PostgreSQL queries

Reviewer: ```suggestion
### Schemaful indexing with PostgreSQL via pnpm
```

> This is the method to store the table information in PostgreSQL, with a table corresponding to each MUD table.
> Use this method when you want to use SQL queries to view the data.

Reviewer: ```suggestion
The schemaful indexer dynamically creates a PostgreSQL table every time it encounters a new MUD table.
It then decodes the MUD events and stores it in the PostgreSQL table with a matching schema.
This approach doesn't scale well to all tables of a chain, but it is a convenient way to index the tables of a particular MUD World and have access to powerful SQL queries on its data.
```

1. Start a `World` to index.
   An easy way to do this is to [use a TypeScript template](/templates/typescript/getting-started) in a separate command line window.

1. Install the indexer.

   ```bash copy
   pnpm install @latticexyz/store-indexer@next
   ```

1. Set the environment variables and start the indexer.

   ```bash copy
   export RPC_HTTP_URL=http://127.0.0.1:8545
   export DATABASE_URL=postgres://127.0.0.1/postgres
   pnpm postgres-decoded-indexer
   ```

> 1. To verify the installation, but `psql` and then:

Reviewer: ```suggestion
1. To verify the installation, run `psql` and then:
```

   **Note:** You should do this every time you restart the blockchain.
   Otherwise your index will include data from multiple blockchains, and make no sense.

#### Schemaless indexing with PostgreSQL via Docker

> The command to start the indexer in PostgreSQL mode is `pnpm start:postgres`.

Reviewer: ```suggestion
The command to start the indexer in schemaless PostgreSQL mode is `pnpm start:postgres`.
```

   **Note:** You should do this every time you restart the blockchain.
   Otherwise your index will include data from multiple blockchains, and make no sense.

> #### PostgreSQL as storage

Reviewer: ```suggestion
#### Schemaless indexing with PostgreSQL via Docker
```

The command to start the indexer in PostgreSQL mode is `pnpm start:postgres`.

> This command starts both the indexer and the front end.

Reviewer: ```suggestion
This command starts both the indexer and query frontends.
```

Reviewer: We should add a section for starting only the writer or only the query frontend (starting the frontend separately is necessary to scale reads)

     pnpm start:postgres
   ```

>    If you want to create additional front ends (for reliability and load balancing), use these commands:

Reviewer: ```suggestion
   If you want to create additional query frontends (for reliability and load balancing), use these commands:
```

   ```sh copy
   HOST_PORT=3002
   docker run \
     --platform linux/amd64 \
     -e DATABASE_URL=postgres://host.docker.internal/postgres \
     -p $HOST_PORT:3001  \
     ghcr.io/latticexyz/store-indexer:latest  \
     node dist/bin/postgres-frontend.js
   ```

#### Schemaful indexing with PostgreSQL via Docker

> The command to start the indexer in PostgreSQL mode is `pnpm start:postgres-decoded`.

Reviewer: ```suggestion
The command to start the indexer in schemaful PostgreSQL mode is `pnpm start:postgres-decoded`.
```

     pnpm start:postgres
   ```

> #### PostgreSQL for queries

Reviewer: ```suggestion
#### Schemaful indexing with PostgreSQL via Docker
```

The command to start the indexer in PostgreSQL mode is `pnpm start:postgres-decoded`.

> This command starts both the indexer and the front end.

Reviewer: ```suggestion
This command starts both the indexer and the query frontend.
```

1. The docker instance identifies itself to PostgreSQL as `root`.
   To give this user permissions on the database, enter `psql` and run this command.

   ```sql copy
   CREATE ROLE root SUPERUSER LOGIN;
   ```

   **Note:** This is assuming a database that is isolated from the internet and only used by trusted entities.
   In a production system you will use at least a password as authentication, and limit the user's authority.

1. Start the docker container.
   For example, to index an `anvil` instance running to the host to the database `postgres` on the host, use.

   ```sh copy
   docker run \
     --platform linux/amd64 \
     -e RPC_HTTP_URL=http://host.docker.internal:8545 \
     -e DATABASE_URL=postgres://host.docker.internal/postgres \
     -p 3001:3001  \
     ghcr.io/latticexyz/store-indexer:latest  \
     pnpm start:postgres-decoded
   ```

### Verification

> If you use either SQLite or PostgreSQL with the front-end (using PostgreSQL only as storage), you can run this command to test the indexer.

Reviewer: ```suggestion
If you use either SQLite or PostgreSQL with the query frontend (using PostgreSQL only as storage), you can run this command to test the indexer.
```

</details>

## Some examples

### SQLite to view the tasks from the React template

- Start the indexer.

  ```sh copy
  pnpm install @latticexyz/store-indexer@next
  export RPC_HTTP_URL=http://127.0.0.1:8545
  pnpm sqlite-indexer
  ```

- Read the data.

  ```sh copy
  curl 'http://localhost:3001/trpc/findAll?batch=1&input=%7B%220%22%3A%7B%22json%22%3A%7B%22chainId%22%3A31337%2C%22address%22%3A%220x6e9474e9c83676b9a71133ff96db43e7aa0a4342%22%7D%7D%7D' | \
  jq '.[0].result.data.json.tables' | \
  jq -c '.[] | select (.name=="Tasks")' | \
  jq .records | more
  ```

> ### Read data from SkyStrife season 0

Reviewer: What i had in mind with the sky strife data was to run the indexer on a world with more data to showcase the indexer
